sklearn/tree/_utils.pyx

```diff
@@ -361,14 +362,8 @@ cdef class WeightedPQueue:
         # Resize if capacity not sufficient
         if array_ptr >= self.capacity:
             self.capacity *= 2
-            array = <WeightedPQueueRecord*> realloc(self.array_,
-                                                    self.capacity *
-                                                    sizeof(WeightedPQueueRecord))
-
-            if array == NULL:
-                # no free; __dealloc__ handles that
-                return -1
-            self.array_ = array
```

Comment:
Indeed it can be extended to all the data structures but I first need to confirm if releasing gil in `safe_realloc` is safe... I can't figure out why it was not done so far... @glouppe or @jmschrei can help us figure that out maybe? :) If the current changes look okay, then I can replace all `malloc` calls with `safe_realloc`... There is also yet another line in `tree.pyx` where `safe_realloc` was [explicitly avoided](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/tree/_tree.pyx#L679) because it used to not release gil...

The thing is tree tests are not very extensive as they need to be and I am paranoid that these kind of clean ups extraneous to issue at hand might break user code in subtle ways...

Thx for raising the question!